src/idct.rs

```diff
@@ -25,41 +30,393 @@ pub(crate) fn choose_idct_size(full_size: Dimensions, requested_size: Dimensions
+            Dimensions {
+                width: 1400,
+                height: 1400
+            }
+        ),
+        4
+    );
+
+    assert_eq!(
+        choose_idct_size(
+            Dimensions {
+                width: 5472,
+                height: 3648
+            },
+            Dimensions {
+                width: 5472,
+                height: 3648
+            }
+        ),
+        8
+    );
+    assert_eq!(
+        choose_idct_size(
+            Dimensions {
+                width: 5472,
+                height: 3648
+            },
+            Dimensions {
+                width: 16384,
+                height: 16384
+            }
+        ),
+        8
+    );
+    assert_eq!(
+        choose_idct_size(
+            Dimensions {
+                width: 1,
+                height: 1
+            },
+            Dimensions {
+                width: 65535,
+                height: 65535
+            }
+        ),
+        8
+    );
+    assert_eq!(
+        choose_idct_size(
+            Dimensions {
+                width: 5472,
+                height: 3648
+            },
+            Dimensions {
+                width: 16384,
+                height: 16384
+            }
+        ),
+        8
+    );
 }
 
-pub(crate) fn dequantize_and_idct_block(scale: usize, coefficients: &[i16], quantization_table: &[u16; 64], output_linestride: usize, output: &mut [u8]) {
+pub(crate) fn dequantize_and_idct_block(
+    scale: usize,
+    coefficients: &[i16],
+    quantization_table: &[u16; 64],
+    output_linestride: usize,
+    output: &mut [u8],
+) {
     match scale {
-        8 => dequantize_and_idct_block_8x8(coefficients, quantization_table, output_linestride, output),
-        4 => dequantize_and_idct_block_4x4(coefficients, quantization_table, output_linestride, output),
-        2 => dequantize_and_idct_block_2x2(coefficients, quantization_table, output_linestride, output),
-        1 => dequantize_and_idct_block_1x1(coefficients, quantization_table, output_linestride, output),
+        8 => dequantize_and_idct_block_8x8(
+            coefficients,
+            quantization_table,
+            output_linestride,
+            output,
+        ),
+        4 => dequantize_and_idct_block_4x4(
+            coefficients,
+            quantization_table,
+            output_linestride,
+            output,
+        ),
+        2 => dequantize_and_idct_block_2x2(
+            coefficients,
+            quantization_table,
+            output_linestride,
+            output,
+        ),
+        1 => dequantize_and_idct_block_1x1(
+            coefficients,
+            quantization_table,
+            output_linestride,
+            output,
+        ),
         _ => panic!("Unsupported IDCT scale {}/8", scale),
     }
 }
 
+macro_rules! idct8_ssse3 {
+    ($data:expr) => {
+        let p2 = $data[2];
+        let p3 = $data[6];
+        let p1 = _mm_mulhrs_epi16(_mm_adds_epi16(p2, p3), _mm_set1_epi16(17734)); // 0.5411961
```

Comment:
Can you outline these constants to some common table/struct and explain them? I'm not sure I understand how they are derived other than the ones related to the common factors. I notice that [jpeg-turbo](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/c23672ce52ae53bd846b555439aa0a070b6d2c07/simd/mips/jsimd.c) uses different—related to the standard ones—and fewer constants. Potentially derive the fixpoint representation from the float as a constant, making this easier to check / modify.

Reply:
The general idea here is that x86 has a fast "fixed point multiplication by [-1, 1) constant represented with 15 mantissa bits" operation (FWIW, so does NEON). So the constants here are just the fractional part of the constants in the non-SIMD implementation multiplied by `1<<15` and rounded to the nearest int. Maybe I should put that in a comment ;)

Reply:
Added a comment.